File: src/cpp/implementation/inner_set_0.cpp
#include "inner_set.hpp"
#include "inner_set_0.hpp"

template<typename T>
inner_set_0<T>::inner_set_0() : inner_set<T>()
{
}

template<typename T>
inner_set_0<T>::~inner_set_0()
{

}

template<typename T>
inner_set<T>* inner_set_0<T>::add_element(T* p_element)
{
    return new inner_set_1<T>(p_element);
}

template<typename T>
inner_set<T>* inner_set_0<T>::add_elements(T** p_elements)
{
    //TODO
    return nullptr;
}

template<typename T>
bool inner_set_0<T>::contains_element(T* p_element)
{
    return false;
}

template<typename T>
bool inner_set_0<T>::contains_all_elements(T* p_elements)
{
    return false;
}

template<typename T>
T* inner_set_0<T>::get_element(int p_index)
{
    return nullptr;
}

template<typename T>
inner_set<T>* inner_set_0<T>::remove_element(T* p_element)
{
    return this;
}

template<typename T>
inner_set<T>* inner_set_0<T>::retain_all_element(T** p_elements)
{
    return this;
}

template<typename T>
int inner_set_0<T>::get_size()
{
    return 0;
}

File: src/scala/implementation/InnerSet_0.scala
package scala.implementation

object InnerSet_0{
  def apply[T]: InnerSet_0[T] = new InnerSet_0[T]
}

class InnerSet_0[T] extends AbstractInnerSet[T]{

  override def addElement(p_element: T): InnerSet[T] = InnerSet_1(p_element)

  override def addUnchecked(p_element: T): InnerSet[T] = addElement(p_element)

  override def addAllElements(p_innerSet: InnerSet[T]): InnerSet[T] = p_innerSet

  override def containsElement(p_element: Object): Boolean = false

  override def containsAllElements(p_innerSet: InnerSet[T]): Boolean = false

  override def getElement(p_index: Int): Option[T] = null

  override def removeElement(p_element: Object): InnerSet[T] = this

  override def removeAllElements(p_innerSet: InnerSet[T]): InnerSet[T] = this

  override def retainAllElements(p_innerSet: InnerSet[T]): InnerSet[T] = this

  override def getSize: Int = 0

  override def iterator: InnerSetIterator[T] = new InnerSetIterator[T](this)

  override def copy: Option[InnerSet[T]] = null

  override def clear(unused: Boolean): InnerSet[T] = this

  override def toString: String = "{ }"

  override def addAndPropagate(p_element: T, p_microSet: MicroSet[T]): InnerSet[T] =
  {
    p_microSet.add(p_element)
    InnerSet_1(p_element)
  }

  override def addAllAndPropagate(p_innerSet: InnerSet[T], p_microSet: MicroSet[T]): InnerSet[T] =
  {
    val microSet: MicroSet[T] = MicroSet(p_innerSet)
    p_microSet.addAll(microSet)
    p_innerSet
  }
}
